internal/mesh/internal/controllers/explicitdestinations/controller.go

```diff
@@ -0,0 +1,235 @@
+			rt.Logger.Trace("updating status", "id", dst.GetResource().GetId())
+			_, err = rt.Client.WriteStatus(ctx, &pbresource.WriteStatusRequest{
+				Id:     dst.GetResource().GetId(),
+				Key:    ControllerName,
+				Status: updatedStatus,
+			})
+			if err != nil {
+				rt.Logger.Error("error writing new status", "id", dst.GetResource().GetId(), "error", err)
+				return err
+			}
+		}
+	}
+
+	// If after fetching and validating, we don't have any destinations,
+	// we need to skip reconcile and delete the resource.
+	if len(newComputedDestinationsData.GetDestinations()) == 0 {
+		rt.Logger.Trace("found no destinations associated with this workload")
+
+		if cpc != nil {
+			rt.Logger.Trace("deleting ComputedDestinations")
+			_, err = rt.Client.Delete(ctx, &pbresource.DeleteRequest{Id: req.ID})
+			if err != nil {
+				// If there's an error deleting CD, we want to re-trigger reconcile again.
+				rt.Logger.Error("error deleting ComputedDestinations", "error", err)
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	// Lastly, write the resource.
+	if cpc == nil || !proto.Equal(cpc.GetData(), newComputedDestinationsData) {
+		rt.Logger.Trace("writing new ComputedDestinations")
+
+		// First encode the endpoints data as an Any type.
+		cpcDataAsAny, err := anypb.New(newComputedDestinationsData)
+		if err != nil {
+			rt.Logger.Error("error marshalling latest ComputedDestinations", "error", err)
+			return err
+		}
+
+		_, err = rt.Client.Write(ctx, &pbresource.WriteRequest{
+			Resource: &pbresource.Resource{
+				Id:    req.ID,
+				Owner: workloadID,
+				Data:  cpcDataAsAny,
+			},
+		})
+		if err != nil {
+			rt.Logger.Error("error writing ComputedDestinations", "error", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validate(
+	ctx context.Context,
+	client pbresource.ResourceServiceClient,
+	destinations *types.DecodedDestinations) (bool, *pbresource.Condition) {
+	for _, dest := range destinations.GetData().GetDestinations() {
+		serviceRef := resource.ReferenceToString(dest.DestinationRef)
+
+		// Fetch and validate service.
+		service, err := resource.GetDecodedResource[*pbcatalog.Service](ctx, client, resource.IDFromReference(dest.DestinationRef))
+		if err != nil {
+			return false, ConditionDestinationServiceReadError(serviceRef)
+		}
+		if service == nil {
+			return false, ConditionDestinationServiceNotFound(serviceRef)
+		}
+
+		if !service.GetData().IsMeshEnabled() {
+			return false, ConditionMeshProtocolNotFound(serviceRef)
+		}
+
+		if service.GetData().FindServicePort(dest.DestinationPort).Protocol == pbcatalog.Protocol_PROTOCOL_MESH {
+			return false, ConditionMeshProtocolDestinationPort(serviceRef, dest.DestinationPort)
+		}
+
+		// Fetch and validate computed routes for service.
+		serviceID := resource.IDFromReference(dest.DestinationRef)
+		cr, err := resource.GetDecodedResource[*pbmesh.ComputedRoutes](ctx, client, resource.ReplaceType(pbmesh.ComputedRoutesType, serviceID))
```

Comment:
I believe it's technically a bug that we fetch `ComputedRoutes` in reconcile and don't watch it. It's a minor bug, but can happen when `ComputedRoutes` change.

Reply:
We also aren't watching `Service`.

Reply:
The routes controller will only produce a ComputedRoutes iff :
- service exists
- has a mesh port

the ComputedRoutes resource will have one named port key in the PortedConfigs map for each port in the Service _except_ any mesh ports.

I think you could skip any of the `Service` checks here and just roll with doing it indirectly via `ComputedRoutes`. Then when we only need to add watches for one new type here.

Reply:
I don't think it's a bug because sidecar proxy controller will update the status of this resource if something is invalid, triggering re-reconcile and re-validation again. https://github.com/hashicorp/consul/pull/19083/files#r1347902548.

Technically, this controller doesn't care about these resources changing outside of updating status in case something is invalid. Services and computed routes don't impact the contents of the computed resource.

If you both are ok with that approach, I can just add a comment clarifying why we're not watching these resources. The weird thing is that we are adding a dependency between controllers here, meaning if sidecar proxy controller stops doing these status updates, this controller will have a bug. On the other hand, if this controller also needs to watch services and computed routes, we end up with more watches and reconciles overall for both controllers because sidecar proxy controller will still need to watch both also. 

>I think you could skip any of the Service checks here and just roll with doing it indirectly via ComputedRoutes. Then when we only need to add watches for one new type here.

I think we should still validate it because everything is eventually consistent, and so the validation of computed routes controller did may no longer apply because the state of the cluster has changed.

Reply:
It looks like the validation logic here is very similar to the validation logic in the other PR. It feels like we should either do all the validation logic here or do all the logic in the sidecar proxy controller. Overall, I think it's slightly cleaner to keep all the logic in this controller so controllers that use computed explicit destinations are just consumers and don't have to know about validation logic, but I totally understand your point.

If we go with the cross controller validation approach, we definitely should add cross-controller tests at some point (like you said).

Reply:
>It looks like the validation logic here is very similar to the validation logic in the other PR.  It feels like we should either do all the validation logic here or do all the logic in the sidecar proxy controller. Overall, I think it's slightly cleaner to keep all the logic in this controller so controllers that use computed explicit destinations are just consumers and don't have to know about validation logic

Yeah it is similar. Originally, I was thinking that we can only do validation in this controller and just assume things are "validated" in the sidecar-proxy controller. However, I realized that because we still need to fetch the latest resources (services and routes) in the sidecar proxy controller anyway, we could be in a situation where this controller thought everything is valid, a user did something and by the time sidecar proxy controller fetched them, it's invalid. And so in the sidecar proxy controller we still need to do something in that case, otherwise, we'd just be skipping resources without reporting errors to the user which didn't seem ideal either in this case.

Reply:
We could do some sort of hybrid approach where:
* The explicit destinations controller is responsible for validating destinations and updating statuses.
* Consumers fetch destinations and do validations on them without updating statuses. Maybe they could even use a lightly modified version of this function if we could figure out a good place to put it. This handles eventual consistency issues.

Reply:
On more thought, I realized that the reverse triggering of statuses will end up having issues where we could possibly miss events and end up with outdated resource. So I followed your advice and updated it to watch services and routes. I'll update sidecar proxy controller with the hybrid approach as you suggest.